IGC/Compiler/Optimizer/IntDivConstantReduction.cpp

```diff
@@ -520,13 +545,29 @@ struct IntDivConstantReduction : public FunctionPass
     {
         //////////////////////////////////////////////////
         // C.f. Hacker's Delight 10-8
+#if LLVM_VERSION_MAJOR >= 14
+        UnsignedDivisonByConstantInfo appxRecip = UnsignedDivisonByConstantInfo::get(divisor);
+#else
         APInt::mu appxRecip = divisor.magicu();
+#endif
         //
         const int bitSize = dividend->getType()->getIntegerBitWidth();
         //
         // even divisors can pre-shift the dividend to avoid
         // extra work at the end.
         Value *shiftedDividend = dividend;
+#if LLVM_VERSION_MAJOR >= 14
+        if (appxRecip.IsAdd && !divisor[0]) {
+            unsigned s = divisor.countTrailingZeros();
+            shiftedDividend = B.CreateLShr(shiftedDividend, s);
+            appxRecip = UnsignedDivisonByConstantInfo::get(divisor.lshr(s));
+            IGC_ASSERT_MESSAGE(!appxRecip.IsAdd, "expected to subtract now");
+            IGC_ASSERT_MESSAGE(appxRecip.ShiftAmount < divisor.getBitWidth(), "undefined shift");
+        }
+        //
+        ConstantInt *appxRcp = IGC::getConstantUInt(
+            B, bitSize, appxRecip.Magic.getZExtValue());
+#else
         if (appxRecip.a && !divisor[0]) {
             unsigned s = divisor.countTrailingZeros();
             shiftedDividend = B.CreateLShr(shiftedDividend, s);
```

Comment:
There is basically the same code under `#if` and `#else`. We cannot accept this. If `ShiftAmount` & `IsAdd` are the only differences we need to put this is the wrapper somehow.